src/Interfaces/AbstractPolyhedron_functions.jl

```diff
@@ -1052,14 +1056,55 @@ julia> project(P, [1, 2]) |> constraints_list
 ```
 """
 function project(P::AbstractPolyhedron{N}, block::AbstractVector{Int}) where {N}
-    if constrained_dimensions(P) ⊆ block
-        clist = [HalfSpace(c.a[block], c.b) for c in constraints_list(P)]
+    general_case = false
+
+    # cheap case
+    clist = nothing  # allocate later
+    @inbounds for c in constraints(P)
+        subset_status = 0  # initially 0, set to 1 if subset and -1 if disjoint
+        for i in constrained_dimensions(c)
+            if i in block
+                # case 1: is a subset
+                if subset_status == -1
+                    general_case = true
+                    break
+                elseif subset_status == 0
+                    subset_status = 1
+                end
+            else
+                # case 1: is disjoint
+                if subset_status == 1
+                    general_case = true
+                    break
+                elseif subset_status == 0
+                    subset_status = -1
+                end
+            end
+        end
+        if subset_status == 1
+            # simple projection of half-space
+            hs = HalfSpace(c.a[block], c.b)
+            if clist == nothing
+                clist = [hs]  # get the right type of the constraints
+            else
+                push!(clist, hs)
+            end
+        end
+    end
+
+    # general case
+    if general_case
+        clist = _project_polyhedron(P, block)
+    end
+
+    if isbounded(P)
+         return HPolytope(clist; check_boundedness=false)
     else
-        n = dim(P)
-        M = projection_matrix(block, n, N)
-        lm = linear_map(M, P)
-        clist = constraints_list(lm)
+         return HPolyhedron(clist)
     end
-    T = isbounded(P) ? HPolytope : HPolyhedron
-    return T(clist)
+end
+
+function _project_polyhedron(P::LazySet, block)
+    lm = project(P, block, nothing)
```

Comment:
i know this is an unrelated comment, but i find that calling again `project(P, block, nothing)` to be unclear as what is the code actually doing (the `nothing` is strange), while we could do `constraints_list(linear_map(projection_matrix(block, dim(P), N), P))` without needing to call again `project`.

Reply:
I don't know. `project` gives you a concrete projection. I can remove the `nothing if you want, that part is optional. I prefer to reuse existing code rather than having a copy.

Reply:
we get a stackoverflow by calling again `project`: https://github.com/JuliaReach/LazySets.jl/issues/2539